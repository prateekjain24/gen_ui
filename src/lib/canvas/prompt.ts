import type { CanvasRecipeId } from "./recipes";

const CANVAS_SYSTEM_PROMPT = `You classify onboarding intents into one of four deterministic recipes.
Recipes:
- R1 Explorer Quick Start → Lightweight callout + optional workspace name + ai_hint + checklist. For solo or vague prompts.
- R2 Team Workspace → Workspace name, team size, integrations, teammate invites, admin toggle. For team/collaboration intents (Slack, Jira, invites, team size ≥ 3).
- R3 Client Project → Workspace name, project type client, integrations (GDrive/Slack/Asana), ai_hint for sharing, kickoff checklist. For client/agency/stakeholder language.
- R4 Power/Compliance → Workspace name, admin toggle, audit logging checkbox, access level select, security integrations (Okta/Jira). For governance/security/audit/policy needs.

Always respond with valid minified JSON:
{
  "persona": "explorer|team|power",
  "recipe_id": "R1|R2|R3|R4",
  "intent_tags": ["tag"],
  "confidence": number between 0 and 1,
  "reasoning": "≤120 chars explaining the match"
}

Rules:
- Use persona "power" only when approvals/audit/security/compliance are explicit.
- Use persona "team" for team/client collaboration including invites, stakeholders, agencies, etc.
- Use persona "explorer" as the safe default for solo or unclear prompts.
- Keep intent_tags lowercase snake_case (e.g., "integrations", "invites", "governance", "client", "solo"). Include at most 3 tags.
- If unsure, set confidence ≤ 0.5 and prefer R1.`;

export interface CanvasClassifierContext {
  domainEmail?: string;
  teamSize?: string;
  metadata?: Record<string, unknown>;
}

export interface BuildCanvasPromptArgs extends CanvasClassifierContext {
  message: string;
}

const sanitizeValue = (value: unknown): string => {
  if (value === undefined || value === null) {
    return '';
  }
  if (typeof value === 'string') {
    return value.trim();
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  return JSON.stringify(value);
};

export const CANVAS_CLASSIFIER_SYSTEM_PROMPT = CANVAS_SYSTEM_PROMPT;

export function buildCanvasClassifierPrompt({ message, domainEmail, teamSize, metadata }: BuildCanvasPromptArgs): string {
  const contextLines: string[] = [];

  if (domainEmail) {
    contextLines.push(`Domain email: ${sanitizeValue(domainEmail)}`);
  }

  if (teamSize) {
    contextLines.push(`Team size: ${sanitizeValue(teamSize)}`);
  }

  if (metadata && Object.keys(metadata).length > 0) {
    contextLines.push(`Metadata: ${sanitizeValue(metadata)}`);
  }

  const formattedMessage = message.trim();
  const contextBlock = contextLines.length > 0 ? `\nContext:\n- ${contextLines.join('\n- ')}` : '';

  return `Classify the following Canvas Chat request into a recipe.\n\nUser message:\n"""${formattedMessage}"""${contextBlock}\n\nReturn only the JSON object described in the system instructions.`;
}

export type CanvasClassifierRecipe = CanvasRecipeId;
