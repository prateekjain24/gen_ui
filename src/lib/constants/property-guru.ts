/**
 * PropertyGuru-specific constant lookups shared across extractor, UI, and telemetry.
 */

import type {
  PropertyGuruLifestyleCue,
  PropertyGuruMoveInHorizon,
  PropertyGuruPropertyType,
  PropertyGuruTonePreset,
} from '@/lib/types/property-guru';

export const PROPERTY_GURU_PROPERTY_TYPES: readonly PropertyGuruPropertyType[] = [
  'condo',
  'executive_condo',
  'hdb',
  'landed',
  'serviced_apartment',
  'co_living',
  'other',
] as const;

export const PROPERTY_GURU_LIFESTYLE_CUES: readonly PropertyGuruLifestyleCue[] = [
  'near_mrt',
  'near_schools',
  'childcare_ready',
  'family_friendly',
  'pet_friendly',
  'wheelchair_access',
  'healthcare_access',
  'green_spaces',
  'co_working',
  'serviced_amenities',
  'rental_hotspot',
  'move_in_ready',
  'renovation_friendly',
  'dual_city',
  'community_events',
] as const;

export const PROPERTY_GURU_TONE_PRESETS: readonly PropertyGuruTonePreset[] = [
  'reassuring',
  'data_driven',
  'concierge',
] as const;

export const PROPERTY_GURU_MOVE_IN_HORIZONS: readonly PropertyGuruMoveInHorizon[] = [
  'immediate',
  'three_months',
  'six_months',
  'year_plus',
  'flexible',
  'unspecified',
] as const;
