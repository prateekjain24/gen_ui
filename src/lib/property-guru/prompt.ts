import { z } from 'zod';

import type { PropertyGuruSignals } from '@/lib/types/property-guru';

export const PROPERTY_GURU_PLAN_SYSTEM_PROMPT = `You are a PropertyGuru onboarding planner helping home seekers turn a conversational brief into an actionable discovery plan.

Always respond with JSON that matches the provided schema. Your plan must feel like a concierge advisor who understands the seeker. Reflect the user's tone preference when available.

### Plan structure
1. intent_summary — 1-2 sentences affirming the seeker's goal.
2. essentials — key filters to tune first (location radius, price band, property type/bedrooms). Present as labelled items with short guidance and recommended values.
3. lifestyle_filters — highlight lifestyle toggles worth adjusting (amenities, accessibility, commute, family cues). Explain why each matters.
4. plan_actions — primary and optional secondary actions to continue in PropertyGuru (e.g., view matches, start mortgage check, schedule tour). Include CTA labels and contextual helper copy.
5. micro_copy — reassurance + follow-up language (listing freshness, weekly digest, concierge support).

### Guardrails
- Never invent impossible data; stay within the signals provided.
- Use Singapore context (currency SGD, MRT references, districts).
- Tone must be one of: reassuring, data-driven, concierge.
- Keep each string ≤ 200 characters.
- Prefer bullet-like phrasing separated by semicolons when listing multiple benefits.
- If information is missing, note sensible defaults (e.g., "Budget not specified — starting with SGD 1.0M").

Return JSON only.`;

export const propertyGuruPlanSchema = z.object({
  intent_summary: z.string().min(1),
  essentials: z.object({
    title: z.string().min(1),
    items: z
      .array(
        z.object({
          label: z.string().min(1),
          value: z.string().min(1),
          helper: z.string().min(1),
        })
      )
      .min(1),
  }),
  lifestyle_filters: z.object({
    title: z.string().min(1),
    highlights: z.array(z.string().min(1)).min(1),
    helper: z.string().min(1),
  }),
  plan_actions: z.object({
    primary: z.object({
      label: z.string().min(1),
      description: z.string().min(1),
    }),
    secondary: z
      .array(
        z.object({
          label: z.string().min(1),
          description: z.string().min(1),
        })
      )
      .default([]),
  }),
  micro_copy: z.object({
    reassurance: z.string().min(1),
    follow_up: z.string().min(1),
  }),
});

export type PropertyGuruPlanTemplate = z.infer<typeof propertyGuruPlanSchema>;

export const buildPropertyGuruPlanPrompt = ({
  prompt,
  signals,
}: {
  prompt: string;
  signals: PropertyGuruSignals;
}): string => {
  const summary = {
    prompt,
    signals,
  };

  return `Context: ${JSON.stringify(summary, null, 2)}\n\nReturn a JSON object that matches the PropertyGuru schema.`;
};

export const createDefaultPropertyGuruPlan = (): PropertyGuruPlanTemplate => ({
  intent_summary: 'Let\'s refine your home search together.',
  essentials: {
    title: 'Tune the essentials',
    items: [
      {
        label: 'Location radius',
        value: '5km around your preferred neighbourhood',
        helper: 'Adjust the slider to widen or tighten nearby listings.',
      },
      {
        label: 'Price band',
        value: 'SGD 1.0M - 1.2M',
        helper: 'Set a range that keeps monthly payments comfortable.',
      },
    ],
  },
  lifestyle_filters: {
    title: 'Lifestyle filters',
    highlights: ['Commute to MRT', 'Nearby schools', 'Pet-friendly options'],
    helper: 'Toggle the priorities that matter; we will reshuffle matches instantly.',
  },
  plan_actions: {
    primary: {
      label: 'Preview curated listings',
      description: 'Opens a shortlist with filters applied.',
    },
    secondary: [
      {
        label: 'Start mortgage check',
        description: 'See monthly repayments and financing options.',
      },
    ],
  },
  micro_copy: {
    reassurance: 'Listings refresh daily so you don\'t miss new launches.',
    follow_up: 'We will email a weekly digest and concierge tips.',
  },
});
