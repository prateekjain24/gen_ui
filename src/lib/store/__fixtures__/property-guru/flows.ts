import type { PropertyGuruPlanTemplate } from '@/lib/property-guru/prompt';
import type { PropertyGuruLifestyleCue, PropertyGuruTonePreset } from '@/lib/types/property-guru';

interface PropertyGuruFlowExpectations {
  location: string;
  propertyType?: string;
  bedrooms?: number;
  lifestyleIncludes?: PropertyGuruLifestyleCue[];
  tone?: PropertyGuruTonePreset;
  nextSteps: string[];
  copyTone: PropertyGuruTonePreset;
  primaryCta: string;
  secondaryCtas?: string[];
  defaultsApplied?: string[];
}

export interface PropertyGuruFlowFixture {
  id: string;
  prompt: string;
  plan: PropertyGuruPlanTemplate;
  expectations: PropertyGuruFlowExpectations;
}

export const PROPERTY_GURU_FLOW_FIXTURES: PropertyGuruFlowFixture[] = [
  {
    id: 'twins_near_tampines',
    prompt:
      "Expecting twins in 4 months, want a 3BR condo near Tampines MRT, budget around 1.4M but I'll stretch for turnkey. Need good preschool options and low-renovation.",
    plan: {
      intent_summary: 'Focused search: 3BR turnkey condo near Tampines MRT within ~SGD 1.4M.',
      essentials: {
        title: 'Tune the essentials first',
        items: [
          {
            label: 'Location radius',
            value: '1.5km around Tampines MRT',
            helper: 'Keeps you within walking distance and near childcare hubs.',
          },
          {
            label: 'Price band',
            value: 'SGD 1.2M – 1.4M (stretch +5% for turnkey units)',
            helper: 'Set alerts at SGD 1.45M so turnkey options surface quickly.',
          },
          {
            label: 'Bedrooms & layout',
            value: '3 bedrooms, turnkey finish',
            helper: 'Filter for renovated family layouts with dual nursery potential.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Lifestyle filters to toggle',
        highlights: [
          'Childcare-ready developments',
          'Family-friendly facilities',
          'Low-renovation inventory',
        ],
        helper: 'Prioritise projects near preschools with playgrounds and service corridors.',
      },
      plan_actions: {
        primary: {
          label: 'Preview curated Tampines listings',
          description: 'Opens a shortlist with MRT proximity and school filters applied.',
        },
        secondary: [
          {
            label: 'Book PropertyGuru Finance consult',
            description: 'Align financing buffers for turnkey units and twins arrival timeline.',
          },
          {
            label: 'Schedule neighbourhood tour',
            description: 'Compare childcare options and transport during a guided visit.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'Listings refresh daily; concierge flags new turnkey options instantly.',
        follow_up: 'Saved searches send weekly MRT-adjacent family picks plus preschool intel.',
      },
    },
    expectations: {
      location: 'Tampines',
      propertyType: 'condo',
      bedrooms: 3,
      lifestyleIncludes: ['near_mrt', 'childcare_ready', 'family_friendly', 'move_in_ready'],
      nextSteps: ['viewListings', 'mortgagePreCheck', 'scheduleTour'],
      copyTone: 'reassuring',
      primaryCta: 'Preview curated Tampines listings',
      secondaryCtas: ['Book PropertyGuru Finance consult', 'Schedule neighbourhood tour'],
    },
  },
  {
    id: 'split_city_worker',
    prompt:
      'Splitting time between Singapore and KL, need a 1BR serviced apartment along the Thomson-East Coast TEL line, want co-working nearby and flexible lease; budget SGD 3.2K monthly.',
    plan: {
      intent_summary: 'Balance your split-city lifestyle with serviced TEL options.',
      essentials: {
        title: 'Lock these in first',
        items: [
          {
            label: 'TEL anchor',
            value: 'Within 1km of Thomson-East Coast Line stations',
            helper: 'Stay near direct airport links for quick KL runs.',
          },
          {
            label: 'Monthly lease budget',
            value: 'SGD 3,200 serviced apartment',
            helper: 'Keep rent aligned with flexible stay expectations.',
          },
          {
            label: 'Layout',
            value: '1 bedroom serviced apartment',
            helper: 'Ensure concierge, housekeeping, and on-demand amenities.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Remote work essentials',
        highlights: ['Co-working access', 'Serviced amenities', 'Flexible lease terms'],
        helper: 'Toggle extras when you need concierge support or hot desk passes.',
      },
      plan_actions: {
        primary: {
          label: 'View serviced apartments near TEL',
          description: 'See TEL-adjacent units with concierge and co-working perks.',
        },
        secondary: [
          {
            label: 'Arrange remote video tour',
            description: 'Line up virtual walkthroughs while you are in KL.',
          },
          {
            label: 'Book PropertyGuru concierge consult',
            description: 'Get help balancing flexible lease terms across cities.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We will flag concierge-ready TEL units as soon as they refresh.',
        follow_up: 'Expect mid-week digests and concierge nudges when new serviced stock appears.',
      },
    },
    expectations: {
      location: 'Thomson-east Coast Line',
      propertyType: 'serviced_apartment',
      bedrooms: 1,
      lifestyleIncludes: ['dual_city', 'co_working', 'serviced_amenities'],
      nextSteps: ['viewListings', 'scheduleTour', 'expertConsult'],
      copyTone: 'reassuring',
      primaryCta: 'View serviced apartments near TEL',
      secondaryCtas: ['Arrange remote video tour', 'Book PropertyGuru concierge consult'],
      defaultsApplied: ['filters.districts'],
    },
  },
  {
    id: 'yield_investor',
    prompt:
      'Looking for a 2BR resale condo in Queenstown under 1.3M, strong rental demand, okay with minor reno, want data on yield and past transactions.',
    plan: {
      intent_summary: 'Spot 2BR yield winners in Queenstown under SGD 1.3M.',
      essentials: {
        title: 'Dial in the fundamentals',
        items: [
          {
            label: 'District focus',
            value: 'District 3 – Queenstown resale stock',
            helper: 'Locks in proven rental demand around MRT and business parks.',
          },
          {
            label: 'Price ceiling',
            value: 'SGD 1.25M – 1.3M',
            helper: 'Keep listings inside your yield target while leaving room for fees.',
          },
          {
            label: 'Layout',
            value: '2 bedroom resale condo',
            helper: 'Choose efficient floorplans tenants favour.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Investor toggles',
        highlights: ['Rental hotspot indicators', 'Reno-friendly layouts', 'Yield comparables'],
        helper: 'Switch on reports and recent transactions while you shortlist.',
      },
      plan_actions: {
        primary: {
          label: 'Review high-yield Queenstown listings',
          description: 'Shortlist data-backed units with tenant demand signals.',
        },
        secondary: [
          {
            label: 'Download rental transaction digest',
            description: 'Export recent transactions and yield benchmarks.',
          },
          {
            label: 'Book investor strategy consult',
            description: 'Walk through financing and renovation math with specialists.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We surface units where yield forecasts and tenant demand stay strong.',
        follow_up: 'Receive new comparables and rent trend alerts every week.',
      },
    },
    expectations: {
      location: 'Queenstown',
      propertyType: 'condo',
      bedrooms: 2,
      lifestyleIncludes: ['rental_hotspot'],
      tone: 'data_driven',
      nextSteps: ['viewListings', 'expertConsult'],
      copyTone: 'data_driven',
      primaryCta: 'Review high-yield Queenstown listings',
      secondaryCtas: ['Download rental transaction digest', 'Book investor strategy consult'],
    },
  },
  {
    id: 'multigenerational_accessibility',
    prompt:
      'Retiring parents moving in, need 4BR near Bishan, lift access, wheelchair friendly, near hospitals and parks; budget 2.3M; we can wait 6-9 months.',
    plan: {
      intent_summary: 'Plan a Bishan move-in that keeps the whole family comfortable.',
      essentials: {
        title: 'Lock these in early',
        items: [
          {
            label: 'Location radius',
            value: '3km around Bishan with hospital access',
            helper: 'Keeps you minutes from healthcare and park connectors.',
          },
          {
            label: 'Layout & access',
            value: '4 bedroom homes with lift access',
            helper: 'Filter for barrier-free entries and elder-friendly blocks.',
          },
          {
            label: 'Budget',
            value: 'Up to SGD 2.3M with buffer for improvements',
            helper: 'Leaves room for accessibility retrofits if needed.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Comfort toggles',
        highlights: ['Wheelchair-friendly routes', 'Hospitals within 10 minutes', 'Park connectors & greenery'],
        helper: 'Keep wellness cues on so the plan balances care and downtime.',
      },
      plan_actions: {
        primary: {
          label: 'View accessibility-ready Bishan homes',
          description: 'Opens listings vetted for lift access and elder-friendly layouts.',
        },
        secondary: [
          {
            label: 'Schedule hospital-nearby tour',
            description: 'Plot a visit that checks commute times to clinics.',
          },
          {
            label: 'Book accessibility renovation consult',
            description: 'Line up pros for grab bars or layout tweaks before move-in.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We emphasise accessibility and calm routes so parents settle in comfortably.',
        follow_up: 'Expect fortnightly updates while you wait for the right listing.',
      },
    },
    expectations: {
      location: 'Bishan',
      bedrooms: 4,
      lifestyleIncludes: ['wheelchair_access', 'healthcare_access', 'green_spaces'],
      nextSteps: ['viewListings', 'scheduleTour', 'expertConsult'],
      copyTone: 'reassuring',
      primaryCta: 'View accessibility-ready Bishan homes',
      secondaryCtas: ['Schedule hospital-nearby tour', 'Book accessibility renovation consult'],
    },
  },
  {
    id: 'first_time_ec',
    prompt:
      'First time buying, looking at executive condos (EC) in Punggol around 1M, not pre-approved, two kids in primary school, want to stay near primary schools, want grants guidance and weekly listing digests.',
    plan: {
      intent_summary: 'Guide your first EC purchase in Punggol with grants in mind.',
      essentials: {
        title: 'Start with these',
        items: [
          {
            label: 'Location focus',
            value: 'Punggol ECs near schools',
            helper: 'Stay close to primary schools and community hubs.',
          },
          {
            label: 'Budget band',
            value: 'Around SGD 1.0M with grant support',
            helper: 'Keep monthly repayments friendly while grants process.',
          },
          {
            label: 'Financing prep',
            value: 'Line up HDB and PropertyGuru Finance guidance',
            helper: 'Know your grants and approval path before locking a unit.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Family priorities',
        highlights: ['Nearby primary schools', 'Family amenities', 'Weekly digest ready'],
        helper: 'Keep these toggled so the plan nudges you with kid-friendly picks.',
      },
      plan_actions: {
        primary: {
          label: 'Save your EC shortlist',
          description: 'Create a shortlist that refreshes as new launches appear.',
        },
        secondary: [
          {
            label: 'Start HDB grant checklist',
            description: 'Work through grant eligibility with guided steps.',
          },
          {
            label: 'Book PropertyGuru mortgage chat',
            description: 'Get pre-approval coaching before you commit.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We will nudge you on grants and eligibility as you progress.',
        follow_up: 'Weekly digests keep new EC options top of mind.',
      },
    },
    expectations: {
      location: 'Punggol',
      propertyType: 'executive_condo',
      lifestyleIncludes: ['near_schools', 'family_friendly'],
      nextSteps: ['savedSearch', 'startHdbGrantChecklist'],
      copyTone: 'reassuring',
      primaryCta: 'Save your EC shortlist',
      secondaryCtas: ['Start HDB grant checklist', 'Book PropertyGuru mortgage chat'],
    },
  },
];
