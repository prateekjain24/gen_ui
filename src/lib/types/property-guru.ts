/**
 * PropertyGuru seeker signal definitions
 *
 * Shared contract describing the typed signals the PropertyGuru preset
 * expects. This keeps extractors, prompt engineering, UI, and telemetry
 * aligned so new work avoids ad hoc string keys.
 */

export type PropertyGuruTonePreset = 'reassuring' | 'data_driven' | 'concierge';

export type PropertyGuruPropertyType =
  | 'condo'
  | 'executive_condo'
  | 'hdb'
  | 'landed'
  | 'serviced_apartment'
  | 'co_living'
  | 'other';

export type PropertyGuruLifestyleCue =
  | 'near_mrt'
  | 'near_schools'
  | 'childcare_ready'
  | 'family_friendly'
  | 'pet_friendly'
  | 'wheelchair_access'
  | 'healthcare_access'
  | 'green_spaces'
  | 'co_working'
  | 'serviced_amenities'
  | 'rental_hotspot'
  | 'move_in_ready'
  | 'renovation_friendly'
  | 'dual_city'
  | 'community_events';

export type PropertyGuruFinanceReadiness =
  | 'pre_approved'
  | 'exploring_options'
  | 'needs_guidance'
  | 'cash_purchase'
  | 'unknown';

export type PropertyGuruMoveInHorizon =
  | 'immediate'
  | 'three_months'
  | 'six_months'
  | 'year_plus'
  | 'flexible'
  | 'unspecified';

export type PropertyGuruBedroomCount = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export interface PropertyGuruLocationFocus {
  /** Primary area, neighbourhood, or planning region name */
  primaryArea: string;
  /** Optional secondary areas to consider */
  secondaryAreas?: string[];
  /** Specific landmarks or transport anchors (e.g. MRT lines) */
  anchorPoints?: string[];
  /** Preferred search radius in kilometres */
  radiusKm?: number;
  /** Optional PropertyGuru district identifiers */
  districtIds?: number[];
}

export interface PropertyGuruPriceBand {
  /** Minimum budget in currency units */
  min?: number;
  /** Maximum comfortable budget */
  max?: number;
  /** Currency code, defaulting to SGD */
  currency: 'SGD';
  /** Whether seeker is open to stretching beyond max (~10%) */
  stretchOk?: boolean;
}

export interface PropertyGuruSignals {
  location: PropertyGuruLocationFocus;
  price: PropertyGuruPriceBand;
  propertyType: PropertyGuruPropertyType;
  bedrooms: PropertyGuruBedroomCount;
  moveInHorizon: PropertyGuruMoveInHorizon;
  lifestyle: PropertyGuruLifestyleCue[];
  financeReadiness: PropertyGuruFinanceReadiness;
  tonePreference: PropertyGuruTonePreset;
}

export type PropertyGuruSignalsPartial = Partial<PropertyGuruSignals>;
