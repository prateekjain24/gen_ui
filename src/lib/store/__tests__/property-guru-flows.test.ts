import { propertyGuruPlanSchema } from '@/lib/property-guru/prompt';
import { PROPERTY_GURU_FLOW_FIXTURES } from '@/lib/store/__fixtures__/property-guru/flows';
import { mapPropertyGuruPlanToSearchPayload } from '@/lib/utils/property-guru-plan-mapper';
import { extractPropertyGuruSignals } from '@/lib/utils/property-guru-signals';

describe('PropertyGuru flow playbooks', () => {
  it.each(PROPERTY_GURU_FLOW_FIXTURES)('%s scenario stays coherent', fixture => {
    const { signals } = extractPropertyGuruSignals(fixture.prompt);

    expect(signals.location.primaryArea).toBe(fixture.expectations.location);

    if (fixture.expectations.propertyType) {
      expect(signals.propertyType).toBe(fixture.expectations.propertyType);
    }

    if (typeof fixture.expectations.bedrooms === 'number') {
      expect(signals.bedrooms).toBe(fixture.expectations.bedrooms);
    }

    if (fixture.expectations.lifestyleIncludes) {
      fixture.expectations.lifestyleIncludes.forEach(cue => {
        expect(signals.lifestyle).toEqual(expect.arrayContaining([cue]));
      });
    }

    if (fixture.expectations.tone) {
      expect(signals.tonePreference).toBe(fixture.expectations.tone);
    }

    const plan = propertyGuruPlanSchema.parse(fixture.plan);
    const { payload, defaultsApplied } = mapPropertyGuruPlanToSearchPayload({ plan, signals });

    expect(plan.plan_actions.primary.label).toBe(fixture.expectations.primaryCta);

    if (fixture.expectations.secondaryCtas) {
      const labels = plan.plan_actions.secondary.map(action => action.label);
      fixture.expectations.secondaryCtas.forEach(label => {
        expect(labels).toEqual(expect.arrayContaining([label]));
      });
    }

    expect(payload.nextSteps).toEqual(expect.arrayContaining(fixture.expectations.nextSteps));
    expect(payload.copy.tone).toBe(fixture.expectations.copyTone);

    if (fixture.expectations.defaultsApplied) {
      expect(defaultsApplied).toEqual(expect.arrayContaining(fixture.expectations.defaultsApplied));
    } else {
      expect(defaultsApplied).toEqual([]);
    }

    expect({
      filters: payload.filters,
      highlights: payload.highlights,
      nextSteps: payload.nextSteps,
      copy: payload.copy,
    }).toMatchSnapshot();
  });
});
