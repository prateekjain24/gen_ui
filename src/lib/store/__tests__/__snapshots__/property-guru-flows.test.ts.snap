// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PropertyGuru flow playbooks {\\n  id: 'first_time_ec',\\n  prompt: 'First time buying, looking at executive condos (EC) in Punggol around 1M, not pre-approved, two kids in primary school, want to stay near primary schools, want grants guidance and weekly listing digests.',\\n  plan: [Object],\\n  expectations: [Object]\\n} scenario stays coherent 1`] = `
{
  "copy": {
    "followUp": "Weekly digests keep new EC options top of mind.",
    "hero": "Guide your first EC purchase in Punggol with grants in mind.",
    "reassurance": "We will nudge you on grants and eligibility as you progress.",
    "tone": "reassuring",
  },
  "filters": {
    "area": "Punggol",
    "bedrooms": 3,
    "districts": [
      19,
    ],
    "lifestyle": [
      "near_schools",
      "family_friendly",
    ],
    "maxPrice": 1000000,
    "minPrice": 1000000,
    "propertyType": "executive_condo",
    "radiusKm": 5,
  },
  "highlights": [
    "Nearby primary schools",
    "Family amenities",
    "Weekly digest ready",
    "Near schools",
    "Family-friendly",
  ],
  "nextSteps": [
    "savedSearch",
    "startHdbGrantChecklist",
  ],
}
`;

exports[`PropertyGuru flow playbooks {\\n  id: 'multigenerational_accessibility',\\n  prompt: 'Retiring parents moving in, need 4BR near Bishan, lift access, wheelchair friendly, near hospitals and parks; budget 2.3M; we can wait 6-9 months.',\\n  plan: [Object],\\n  expectations: [Object]\\n} scenario stays coherent 1`] = `
{
  "copy": {
    "followUp": "Expect fortnightly updates while you wait for the right listing.",
    "hero": "Plan a Bishan move-in that keeps the whole family comfortable.",
    "reassurance": "We emphasise accessibility and calm routes so parents settle in comfortably.",
    "tone": "reassuring",
  },
  "filters": {
    "area": "Bishan",
    "bedrooms": 4,
    "districts": [
      20,
    ],
    "lifestyle": [
      "wheelchair_access",
      "healthcare_access",
      "green_spaces",
      "rental_hotspot",
      "family_friendly",
    ],
    "maxPrice": 2300000,
    "minPrice": 2300000,
    "propertyType": "condo",
    "radiusKm": 5,
  },
  "highlights": [
    "Wheelchair-friendly routes",
    "Hospitals within 10 minutes",
    "Park connectors & greenery",
    "Wheelchair access",
    "Healthcare access",
    "Green spaces",
    "Rental hotspot",
    "Family-friendly",
  ],
  "nextSteps": [
    "viewListings",
    "scheduleTour",
    "expertConsult",
  ],
}
`;

exports[`PropertyGuru flow playbooks {\\n  id: 'split_city_worker',\\n  prompt: 'Splitting time between Singapore and KL, need a 1BR serviced apartment along the Thomson-East Coast TEL line, want co-working nearby and flexible lease; budget SGD 3.2K monthly.',\\n  plan: [Object],\\n  expectations: [Object]\\n} scenario stays coherent 1`] = `
{
  "copy": {
    "followUp": "Expect mid-week digests and concierge nudges when new serviced stock appears.",
    "hero": "Balance your split-city lifestyle with serviced TEL options.",
    "reassurance": "We will flag concierge-ready TEL units as soon as they refresh.",
    "tone": "reassuring",
  },
  "filters": {
    "area": "Thomson-east Coast Line",
    "bedrooms": 1,
    "districts": [],
    "lifestyle": [
      "co_working",
      "serviced_amenities",
      "dual_city",
    ],
    "maxPrice": 3200,
    "minPrice": 3200,
    "propertyType": "serviced_apartment",
    "radiusKm": 5,
  },
  "highlights": [
    "Co-working access",
    "Serviced amenities",
    "Flexible lease terms",
    "Co working",
    "Dual city",
  ],
  "nextSteps": [
    "viewListings",
    "scheduleTour",
    "expertConsult",
  ],
}
`;

exports[`PropertyGuru flow playbooks {\\n  id: 'twins_near_tampines',\\n  prompt: "Expecting twins in 4 months, want a 3BR condo near Tampines MRT, budget around 1.4M but I'll stretch for turnkey. Need good preschool options and low-renovation.",\\n  plan: [Object],\\n  expectations: [Object]\\n} scenario stays coherent 1`] = `
{
  "copy": {
    "followUp": "Saved searches send weekly MRT-adjacent family picks plus preschool intel.",
    "hero": "Focused search: 3BR turnkey condo near Tampines MRT within ~SGD 1.4M.",
    "reassurance": "Listings refresh daily; concierge flags new turnkey options instantly.",
    "tone": "reassuring",
  },
  "filters": {
    "area": "Tampines",
    "bedrooms": 3,
    "districts": [
      18,
    ],
    "lifestyle": [
      "near_mrt",
      "near_schools",
      "childcare_ready",
      "family_friendly",
      "move_in_ready",
      "renovation_friendly",
    ],
    "maxPrice": 1540000,
    "minPrice": 1400000,
    "propertyType": "condo",
    "radiusKm": 1.5,
  },
  "highlights": [
    "Childcare-ready developments",
    "Family-friendly facilities",
    "Low-renovation inventory",
    "Near MRT",
    "Near schools",
    "Childcare ready",
    "Family-friendly",
    "Move-in ready",
    "Renovation friendly",
  ],
  "nextSteps": [
    "viewListings",
    "mortgagePreCheck",
    "scheduleTour",
  ],
}
`;

exports[`PropertyGuru flow playbooks {\\n  id: 'yield_investor',\\n  prompt: 'Looking for a 2BR resale condo in Queenstown under 1.3M, strong rental demand, okay with minor reno, want data on yield and past transactions.',\\n  plan: [Object],\\n  expectations: [Object]\\n} scenario stays coherent 1`] = `
{
  "copy": {
    "followUp": "Receive new comparables and rent trend alerts every week.",
    "hero": "Spot 2BR yield winners in Queenstown under SGD 1.3M.",
    "reassurance": "We surface units where yield forecasts and tenant demand stay strong.",
    "tone": "data_driven",
  },
  "filters": {
    "area": "Queenstown",
    "bedrooms": 2,
    "districts": [
      3,
    ],
    "lifestyle": [
      "rental_hotspot",
    ],
    "maxPrice": 1300000,
    "minPrice": 1300000,
    "propertyType": "condo",
    "radiusKm": 5,
  },
  "highlights": [
    "Rental hotspot indicators",
    "Reno-friendly layouts",
    "Yield comparables",
    "Rental hotspot",
  ],
  "nextSteps": [
    "viewListings",
    "savedSearch",
    "expertConsult",
  ],
}
`;
